[VM] Add whisper feature extractor to Relax VM

LeshengJin opened the pull request:
This pr brings Whisper's feature extractor to Relax VM. The feature extractor extracts mel-filter bank features from raw speech data. The deviation between the current implementation and huggingface's `WhisperFeatureExtractor._np_extract_fbank_features` is less than 1e-4.

tqchen committed:
[MLC][CI] Do not upstream

MLC local ci setup.

MasterJH5574 committed:
[MLC][CI] Do not upstream - Win/Mac Building CI (#137)

This PR adds CI for Windows and macOS building, which may take 90-100
mins.

Co-authored-by: Siyuan Feng <Hzfengsy@sjtu.edu.cn>

jinhongyii committed:
[Distributed] DTensor Data Structure and Parser/Printer Support (#139)

This PR implements basic DTensor structure, and its printer/parser support

MasterJH5574 committed:
[TIR] Blockize keeping T.init in inner block when outer block does no… (#172)

… reduction

This PR modifies the behavior of tir.Schedule.blockize, so that when the
outer block after blockization does no effective reduction, the T.init
part will be kept in the inner block.

NOTE: unit tests regarding blockize may fail due to structural
inequality.

Update to resolve merge.

Co-authored-by: Bohan Hou <spectrometerh@gmail.com>

Co-authored-by: spectrometerHBH <spectrometerh@gmail.com>

jinhongyii committed:
[Dynamic] M2 for S3: Compute Inline (#173)

spectrometerHBH committed:
[Dynamic] M2 for cache_read/write (S2) (#179)

jinhongyii committed:
[Dynamic] M2 for S4: reverse compute inline (#176)

Problem after S3's fix:
 - cannot get producer outside current scope
 - cannot handle dynamic shape in ReverseComputeInliner

Fix:
 - try to get leaf producer block outside current scope
- ignore non-index var when substituting in ReverseComputeInliner's
substituter

junrushao committed:
[Dynamic] M2 for Pad-Einsum (S0) (#181)

MasterJH5574 committed:
[HotFix] Rename BufferReplacer to avoid linking issue (#185)

Both PadEinsum and cross-thread reduction lowering is using
BufferReplacer as a symbol, in which case the linker links in a wrong
way and causes very severe and hard-to-find error.

This PR renames the buffer replacer on both sides.

MasterJH5574 committed:
[Fix][TIR][Schedule] Reverse-Compute-Inline for dyn case (#186)

This PR fixes a bug of recent enhancement in #176.

Credit goes to Bohan.

Co-authored-by: Bohan Hou <spectrometerh@gmail.com>

spectrometerHBH committed:
[Fix] Fix the Maven pom.xml for OS X arm64 tvm4j build (#201)

SiriusNEO committed:
[Training] Bind TE Gradient Func to TIR Name (#195)

This PR provides an interface to let user bind a manually written TE
gradient function to a TIR function name. This binding information will
be stored as an attribute of the module (with attr key
`te_grad_bind_handler`).
When the gradient pass meet a `call_tir`, it will try to search whether
there is a TE gradient function bound to this name. If so, it will call
the handler (i.e. A wrapped function to execute
`bb.emit_te(te_grad_func, ...)`) and get the emitted var. Then the
partials will be set to `var[0], var[1], ...` and we can do the adjoints
propagation.

SiriusNEO committed:
[Fix][Training] Fix Bind TE Grad and Handle ExternOp in Gradient Pass (#207)

This PR contains two quick fixes for current gradient mechanism. First,
it fixes the type of the attribute `te_grad_bind_handler` (use a cast
from dict to tvm Map). Second, we ignore the ExternOp when
`VisitingBinding` in the Gradient pass.

MasterJH5574 committed:
[CMAKE] Add vulkan header Android (#208)

This PR enhances the vulkan search for NDK.

Authored by TQ.

Co-authored-by: tqchen <tqchenml@gmail.com>

Ubospica committed:
[Relax][Training] Update on registering te gradient functions (#209)

This PR registers te gradient functions as global functions. Previously
we will register them as IRModule attributes, but that will lead to
segment fault when program exits due to circular reference.

Current workflow is as follows:
```
# register te gradient function
@register_te_gradient("f_mul_grad")
def f_mul_grad(output_grad: te.Tensor, src1: te.Tensor, src2: te.Tensor, k: int):
    # returns a list of te tensors, representing gradients w.r.t. src1, src2
    # k is a constant parameter
    ...

# irmodule definition
@I.ir_module
class Module:
    @T.prim_func
    def f_mul(A, B, result):
        ...

    @R.function
    def main(a, b):
        cls = Module
        with R.dataflow():
            lv = R.call_tir(cls.f_mul, (a, b), te_grad_name="f_mul_grad", te_grad_kwargs={"k": 1})
            gv = R.output(lv)
        return gv
```

It's worth to note this PR defines an attribute to the call_tir node:
```
struct CallTIRAttrs : public tvm::AttrsNode<CallTIRAttrs> {
  Optional<String> te_grad_name;
  Map<String, ObjectRef> te_grad_kwargs;

  TVM_DECLARE_ATTRS(CallTIRAttrs, "relax.attrs.CallTIRAttrs") {
    TVM_ATTR_FIELD(te_grad_name)
        .describe("The name of the te gradient function associated with this call_tir node.");
    TVM_ATTR_FIELD(te_grad_kwargs)
        .describe("The keyword arguments passed to the te gradient function.");
  }
};  // struct CallTIRAttrs
```

spectrometerHBH committed:
[Target] fix opencl target kind (#210)

spectrometerHBH committed:
Add ndk static build (#211)

MasterJH5574 committed:
Fix NDK static build (#213)

yelite committed:
Fix incorrect bias stride in matmul cutlass offload  (#212)

This PR makes the cutlass codegen use the correct bias stride when bias
has more than 2 dimensions. For example, if the input bias has shape (1,
n, 4096), the original code will set `ldc` to 0, which produces
incorrect result.

cc @vinx13 @masahi

junrushao committed:
rm -rf build/

Ubospica committed:
[Unity][Pass] Checkpointing for Gradient pass (#199)

This PR supports checkpointing for the Gradient pass.

To be specific, in the backward process, we need to use vars computed in
the forward process. Those vars contained in the given checkpoint array,
and the inputs of the function, will be used as is; other vars will be
computed again (this will generate bindings) using the checkpoint vars.

Example:

```
@I.ir_module
class Before:
    @R.function
    def main(x: R.Tensor((3, 3), "float32")):
        R.func_attr({"checkpoint": ["lv2", "gv"]})
        with R.dataflow():
            lv1 = x * R.const(2, "float32")
            lv2 = lv1 * R.const(2, "float32")
            gv = R.sum(lv2)
            R.output(gv)
        return gv

After = relax.transform.Gradient("main")(Before)
After.show(None, False)
```
-->
```
@I.ir_module
class Module:
    @R.function
    def main_adjoint(x: R.Tensor((3, 3), dtype="float32")) -> R.Tuple(R.Tensor((), dtype="float32"), R.Tuple(R.Tensor((3, 3), dtype="float32"))):
        R.func_attr({"checkpoint": ["lv2", "gv"]})
        with R.dataflow():
            lv1: R.Tensor((3, 3), dtype="float32") = R.multiply(x, R.const(2, "float32"))
            lv2: R.Tensor((3, 3), dtype="float32") = R.multiply(lv1, R.const(2, "float32"))
            gv: R.Tensor((), dtype="float32") = R.sum(lv2, axis=None, keepdims=False)
            gv_adjoint: R.Tensor((), dtype="float32") = R.ones(R.shape([]), dtype="float32")
            lv2_adjoint: R.Tensor((3, 3), dtype="float32") = R.broadcast_to(gv_adjoint, R.shape([3, 3]))
            lv1_cp: R.Tensor((3, 3), dtype="float32") = R.multiply(x, R.const(2, "float32"))
            lv1_adjoint: R.Tensor((3, 3), dtype="float32") = R.multiply(lv2_adjoint, R.const(2, "float32"))
            lv: R.Tensor((3, 3), dtype="float32") = R.multiply(lv2_adjoint, lv1_cp)
            x_adjoint: R.Tensor((3, 3), dtype="float32") = R.multiply(lv1_adjoint, R.const(2, "float32"))
            lv1_1: R.Tensor((3, 3), dtype="float32") = R.multiply(lv1_adjoint, x)
            x_adjoint_out: R.Tensor((3, 3), dtype="float32") = x_adjoint
            R.output(gv, x_adjoint_out)
        return (gv, (x_adjoint_out,))

    @R.function
    def main(x: R.Tensor((3, 3), dtype="float32")) -> R.Tensor((), dtype="float32"):
        R.func_attr({"checkpoint": ["lv2", "gv"]})
        with R.dataflow():
            lv1: R.Tensor((3, 3), dtype="float32") = R.multiply(x, R.const(2, "float32"))
            lv2: R.Tensor((3, 3), dtype="float32") = R.multiply(lv1, R.const(2, "float32"))
            gv: R.Tensor((), dtype="float32") = R.sum(lv2, axis=None, keepdims=False)
            R.output(gv)
        return gv
```
lv1_cp is checkpointed

This PR also fixes bugs in te gradients while intergrating with the te
gradient infra.

MasterJH5574 committed:
[Unity] Support clear global memory allocators (#220)

yzh119 committed:
[VM] Add repetition penalty functions to Relax VM  (#219)

MasterJH5574 committed:
[Web] Temporarily work around emcc-ld initial mem too small issue (#221)

junrushao committed:
[CD] Remove constraint on numpy version (#223)

junrushao committed:
Add typing_extensions (#224)

Ubospica committed:
[Transform] Enhance FuncCopier (#225)

Now FuncCopier will provide a var_map member as the relax var mapping
from the old function and the new function.

In FuncCopier we will call SymbolicVarRenewMutator, which also renews
relax Vars. However, previously we did not consider this var mapping
into account, leading that we will provide wrong var_map in FuncCopier.

That leads to bugs in the Gradient pass, whichs uses this member.

This PR will merge the var mappings in FuncCopier and in
SymbolicVarRenewMutator.

junrushao committed:
[PyPI] Add extra directories for JVM/iOS build (#230)

junrushao committed:
[Dynamic] Enhance PadEinsum to support constant indexing (#232)

junrushao committed:
[Fix] Package path to jvm/ and 3rdparty/ (#231)

This PR fixed a bug introduced in
https://github.com/mlc-ai/relax/pull/230, where the directory is
mistakenly specified

junrushao committed:
Fix setup.py when rmtree is executed multiple times (#234)

junrushao committed:
[Pypi] Add src, cmake, CMakeList.txt to Wheels (#237)

tqchen committed:
[CONTRIB][NDK] Update ndk to allow tar libs

tqchen committed:
[TEST] Temp fix to update test gradient in merge

junrushao committed:
[Hotfix] Remove LOG(INFO) from unsupported dtype legalization pass (#240)

Cherry-pick https://github.com/apache/tvm/pull/15015

Co-authored-by: Zihao <expye@outlook.com>

junrushao committed:
Fix static llvm dependency (#242)

LeshengJin committed:
preprocess for whisper

junrushao committed:
[Unity][FIX] add init file to `relax.backend.contrib` (#15023) (#244)

Cherry-pick https://github.com/apache/tvm/pull/15023

LeshengJin committed:
fix

LeshengJin committed:
fix

LeshengJin committed:
skip test if transformers is not available

LeshengJin committed:
fix

LeshengJin committed:
add logits processor

LeshengJin committed:
add runtime support

LeshengJin committed:
add transpose

LeshengJin committed:
fix not max length issue

LeshengJin committed:
runtime bug

yzh119 commented on src/runtime/relax_vm/audio_support.cc:
```diff
@@ -0,0 +1,332 @@
+
+// This is an inplace operation.
+void WhisperProcessLogits(NDArray logits, double cur_len) {
+  std::vector<int> suppress_tokens = {
```
Can we allocate it outside of the function (in global memory space instead of stack)?

yzh119 committed:
[CMake] Find lxml2 for in FindLLVM (#245)

Ubospica committed:
[Fix][Unity][Training] Fix a bug in Gradient tests due to merging (#241)

There was a conflict in the modification about the Gradient on the mlc
and unity branches, which resulted in testing errors. This PR fixes the
issue in tests.

junrushao committed:
[Unity] Add hexp for compute capacity <= 5.2 (#15070) (#246)

upd

Co-authored-by: Zihao Ye <expye@outlook.com>

tqchen committed:
[Unity] Optimize SampleTopPFromProb

This PR optimizes sample top p from prob to reduce its running time
to about 10x or more on M1. The main observation is that we don't
need full sort and can just select top p based on a filtered cap.

junrushao committed:
[Bugfix][CUDA] Fix codegen for hexp for sm >= 52 (#250)

Fix a minor bug introduced by apache#15070

LeshengJin committed:
upd

LeshengJin committed:
Merge remote-tracking branch 'origin' into whisper_support

LeshengJin committed:
upd
